File: leetcode/word_ladder.py
# Given two words (beginWord and endWord), and a dictionary, find the length 
# of shortest transformation sequence from beginWord to endWord, such that:
# Only one letter can be changed at a time
# Each intermediate word must exist in the dictionary
# Time Complexity: O(n)
# Space Complexity: O(n)

from collections import deque

class Solution:
    # @param {string} beginWord
    # @param {string} endWord
    # @param {set<string>} wordDict
    # @return {integer}
    def ladderLength(self, beginWord, endWord, wordDict):
        words, queue = set(wordDict), deque([(beginWord, 1)])
        while queue:
            str, length = queue.popleft()
            for i in xrange(len(str)):
                for c in "abcdefghijklmnopqrstuvwxyz":
                    if c == str[i]: continue
                    new_str = str[:i] + c + str[i + 1:]
                    if new_str == endWord: return length + 1
                    if new_str in words:
                        queue.append((new_str, length + 1))
                        words.remove(new_str)
        return 0

if __name__ == "__main__":
    solution = Solution()
    print solution.ladderLength("hit", "cog", ["hot","dot","dog","lot","log"])

File: leetcode/word_ladder.js
/* Given two words (beginWord and endWord), and a dictionary, find the length 
 * of shortest transformation sequence from beginWord to endWord, such that:
 * Only one letter can be changed at a time
 * Each intermediate word must exist in the dictionary
 * Time Complexity: O(n 26^l), l is the length of each word
 * Space Complexity: O(n)
 */

/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {Set} wordList
 *   Note: wordList is a Set object, see:
 *   https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set
 * @return {number}
 */
var ladderLength = function(beginWord, endWord, wordList) {
    queue = [[beginWord, 1]];
    wordList.add(endWord);
    for (var new_queue = []; queue.length; queue = new_queue) {
        for (new_queue = []; queue.length;) {
            var temp = queue.pop();
            var str = temp[0];
            var length = temp[1];
            for (var i = 0; i < str.length; ++i) {
                var original = str.charAt(i);
                var prefix = str.substring(0, i);
                var suffix = str.substring(i + 1);
                for (var c of "abcdefghijklmnopqrstuvwxyz") {
                    if (c !== original) {
                        var new_str = prefix.concat(c, suffix);
                        if (wordList.has(new_str)) {
                            if (new_str === endWord)
                                return length + 1;
                            new_queue.push([new_str, length + 1]);
                            wordList.delete(new_str);
                        }
                    }
                }
            }
        }
    }
    return 0;
};
